Reapply "[mlir] NFC: fix dependence of (Tensor|Linalg|MemRef|Complex) dialects on LLVM Dialect and LLVM Core in CMake build (#104832)"

christopherbate opened the pull request:
Reapply the commit 43b508566799751aa180f1eaaafc5be693f2f1ae with additional fixes for building with
BUILD_SHARED_LIBS=ON.

aartbik commented:
LGTM for sparse part

matthias-springer commented on mlir/lib/Dialect/MemRef/Transforms/CMakeLists.txt:
```diff
@@ -35,11 +35,12 @@ add_mlir_dialect_library(MLIRMemRefTransforms
   MLIRTransforms
   MLIRValueBoundsOpInterface
-  MLIRVectorDialect
+  MLIRVectorDialect  
```
nit: space

matthias-springer commented on mlir/lib/Conversion/VectorToLLVM/CMakeLists.txt:
```diff
@@ -13,10 +13,11 @@ add_mlir_conversion_library(MLIRVectorToLLVM
   MLIRLLVMCommonConversion
   MLIRLLVMDialect
-  MLIRMemRefDialect
+  MLIRMemRefDialect  
```
nit: space

matthias-springer commented on mlir/lib/Conversion/SCFToControlFlow/SCFToControlFlow.cpp:
```diff
@@ -379,8 +378,8 @@ LogicalResult ForLowering::matchAndRewrite(ForOp forOp,
-                [](auto attr) {
-                  return isa<LLVM::LLVMDialect>(attr.getValue().getDialect());
+                [](NamedAttribute attr) {
+                  return attr.getValue().getDialect().getNamespace() == "llvm";
```
Why is this needed? Can `MLIRSCFToControlFlow` depend on the LLVM dialect?

joker-eph replied:
Using a string here looks like a bit of a hack to me :(

It's pretty unfortunate that we have to choose between a dependency and the feature here though.

christopherbate replied:
> Why is this needed? Can MLIRSCFToControlFlow depend on the LLVM dialect?

Well downstream users that don't use LLVM IR or LLVM MLIR Dialect may want to use this pass.   If downstream users are sensitive to the build/binary cost then they will have to duplicate this pass to eliminate the dependency. 

I guess it's not a huge deal to solve; we can expose the patterns as a separate function and have a callback to customize how attributes are propagated. The current behavior can remain the default and  downstream user can then instantiate a different pass without the LLVM dependency.

christopherbate replied:
removed changes and will follow up with workaround mentioned above

christopherbate committed:
Reapply "[mlir] NFC: fix dependence of (Tensor|Linalg|MemRef|Complex) dialects on LLVM Dialect and LLVM Core in CMake build (#104832)"

Reapply the commit with additional fixes for building with
BUILD_SHARED_LIBS=ON.
